Network/Socket/Types.hsc

```diff
@@ -74,27 +73,24 @@ import Network.Socket.Imports
 -----------------------------------------------------------------------------
 
 -- | Basic type for a socket.
-newtype Socket = Socket (IORef CInt)
+data Socket = Socket (IORef CInt) String
```

Comment:
I see you decided to go with `String` rather than `ShortByteString` or similar. Perhaps slightly more expensive, but easily changed later, without compatibility issues. Are there any module dependency barriers here (circular dependencies or dependencies we want to avoid) to changing this later?  Am I wrong about String being a "more expensive" representation to carry along with the socket?

Of course this is unlikely to matter much...

Reply:
GHC 7.10 uses bytestring-0.10.0.2 which does not provide `ShortByteString` unfortunately.

Reply:
Does it make more sense to carry around an ascii encoded `ByteString` and `pack` it whenever it needs to be shown? Utilizing `String` here means we are introducing a fair amount of indirection and bloat for a `Show` instance.

Reply:
That was my instinct also.  (Though of course ByteString show is `unpack`, but that's just detail).
Perhaps the approach should instead be to store the minimal information needed to avoid having to use `unsafePerformIO` later.  To that end, just the lazy byte string form of the fd is enough.
We could even store an `originalFd` as a `Word32` value, whose different type should be enough to discourage misuse (bypassing the IORef).

```
data Socket = Socket (IORef CInt) Word32
...
 mkSocket :: CInt -> IO Socket
mkSocket fd = do
    ref <- newIORef fd
    let s = Socket ref $ fromIntegral fd
    void $ mkWeakIORef ref $ close s
    return s
```
and then in the `show` instance use that:
```
   show (Socket _ wfd) = "<socket: " ++ show wfd ++ ">"
```
There's no perhaps need to pre-format the whole string (was my suggestion, sorry).  We could keep in mind that Haskell is lazy, and so the current string would I think remain an unevaluated thunk until actually used. Not sure whether the thunk is cheaper than the short string.

Reply:
I changed it to `data Socket = Socket (IORef CInt) CInt`.